### ISSUE TITLE
Does monte_input REALLY replicate the RUN?

### ISSUE THREAD

**Issue by dbankieris** [nasa/trick#710]
A `monte_input` file is generated for each Monte Carlo run by the slave child process. It's meant to be an input file you can pass to `S_main*` to re-run that RUN as it was run by the master. However, `trick.mc_set_enabled(0)` causes `monte_slave_pre` and `monte_slave_post` jobs not to be run. @alexlin0 thinks we might want to run at least the pre-jobs and suggested something along the lines of #705, which Ramtares could also use for their purposes.

@garyt2 @ddj116

**Comment by ddj116** [comment]
Yes please.  In addition to what you have mentioned, any logic behind an `if trick_mc.mc.get_enabled()` gate in the input files will not be executed on subsequent execution of `S_main ... monte_input` for obvious reasons.  Although one could argue the only logic behind that gate should be the setup/creation of dispersions, in practicality there are a couple things users might want to behave differently when running monte carlo such as:

* Logging/DataRecord options and frequencies 
* Custom model-logic during init like correlated dispersion mechanisms.  Typically we set a flag in the input file to tell these models to execute this logic during init.

At the end of the day, in my opinion re-runs of `monte_input` should be guaranteed identical in every way to the original run in the master/slave framework.  Like you mention in #709 this might ultimately require a significant redesign of the monte-carlo architecture.

**Comment by garyt2** [comment]
Got another complication for consideration, related to this and #709, and `monte_master_pre` jobs.  Posting here to record the use-case without making another issue.

If I want to advance a variable with each run, I can schedule a `monte_master_pre` job, something like:
```
my_variable = mc_get_current_run() + offset;
```
then add `my_variable` as a `MonteVarCalculated` so that it gets recorded correctly in the `monte_input` file for later re-runs.

There's a user trap here in that two independent steps are required.  If the second step is missed, the re-runs will not match the initial runs because the re-run will neither pick up the computed value from `monte_input` nor compute that value for itself.  Meanwhile, the first-run will be dispersing a value that is not recorded in `monte_input`.  But that's user-error.  An unfortunate opportunity to accidentally hang oneself, but not a real issue.

More tricky is the case where a sim has to support multiple scenarios, and `my_variable` is only truly variable in one (or some) of those scenarios.  In all the other scenarios where the value of `my_variable` is constant across all runs, the computation must be removed.  That is easy to do for re-runs (simply not add the `MonteVarCalculated`)  but not so much on the first-runs which require that the `monte_master_pre` job be switched off.

So the net sequence for meeting this type of requirement is:

- add a `monte_master_pre` job to compute the value to support those scenarios for which it is variable
- turn the job off very early in the sequence, before it can run at the `monte-master-pre` schedule to support those scenarios for which it is not
- for scenarios that need this to be computed:
  - turn the job back on somewhere between where it was turned off and where it would normally run at the  `monte-master-pre` schedule (is this even possible?)
  - add the variable as a `MonteVarCalculated` instance

An unfortunate consequence of this sequence -- if it is even possible to accomplish --  is that it is necessary to provide individual `monte-master-pre` jobs for each such variable to support independent activation/deactivation of these computations.

I keep coming back to the need for a MonteVar* with executable code (see #702, #706, #709)

**Comment by garyt2** [comment]
Current solution to this predicament is to create a sequence of files each containing one possible set of values for `my_variable`, or create a file on-the-fly for the given situation, then use `MonteVarFile` to extract the value for those scenarios where `my_variable` needs to vary.

This is not an elegant solution.
